File: app/src/main/java/com/supagorn/devpractice/customs/adapter/java/SimpleFragmentPagerAdapter.java
package com.supagorn.devpractice.customs.adapter.java;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import java.util.ArrayList;
import java.util.List;

public class SimpleFragmentPagerAdapter<T> extends FragmentPagerAdapter {
    private int NUM_ITEMS = 0;
    private List<T> items = new ArrayList<>();
    private OnBindViewListener onBindViewListener;

    public SimpleFragmentPagerAdapter(FragmentManager fragmentManager, List<T> items, OnBindViewListener onBindViewListener) {
        super(fragmentManager);
        this.items = items;
        this.NUM_ITEMS = items.size();
        this.onBindViewListener = onBindViewListener;
    }

    // Returns total number of pages
    @Override
    public int getCount() {
        return NUM_ITEMS;
    }

    // Returns the fragment to display for that page
    @Override
    public Fragment getItem(int position) {
        if (onBindViewListener != null) {
            return onBindViewListener.getItem(items.get(position), position);
        } else {
            return null;
        }
    }

    // Returns the page title for the top indicator
    @Override
    public CharSequence getPageTitle(int position) {
        return this.onBindViewListener == null ? "Page " + position : onBindViewListener.getPageTitle(position);
    }


    public interface OnBindViewListener {
        <T> Fragment getItem(T item, int position);

        CharSequence getPageTitle(int position);
    }
}

File: app/src/main/java/com/supagorn/devpractice/customs/adapter/kotlin/SimpleFragmentPagerAdapter.kt
package com.supagorn.devpractice.customs.adapter.kotlin

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter

class SimpleFragmentPagerAdapter<T> : FragmentPagerAdapter {
    private var NUM_ITEMS = 0
    private var items: List<T> = ArrayList()
    private var onBindViewListener: OnBindViewListener? = null

    constructor(
        fragmentManager: FragmentManager,
        items: List<T>,
        onBindViewListener: OnBindViewListener
    ) : super(fragmentManager) {
        this.items = items
        this.NUM_ITEMS = items.size
        this.onBindViewListener = onBindViewListener

    }

    // Returns total number of pages
    override fun getCount(): Int {
        return NUM_ITEMS
    }

    // Returns the fragment to display for that page
    override fun getItem(position: Int): Fragment {
        return if (onBindViewListener != null) {
            onBindViewListener!!.getItem(items[position], position)
        } else {
            Fragment()
        }
    }

    // Returns the page title for the top indicator
    override fun getPageTitle(position: Int): CharSequence? {
        return if (this.onBindViewListener == null) "Page $position" else onBindViewListener!!.getPageTitle(
            position
        )
    }


    interface OnBindViewListener {
        fun <T> getItem(item: T, position: Int): Fragment

        fun getPageTitle(position: Int): CharSequence
    }
}